In [1]:
import pandas as pd
import scipy.optimize as opt
import numpy as np
import statsmodels.api as sm
from statsmodels.iolib.summary2 import summary_col

In [2]:
#Read in the data file PS3_data.dta

est_data = pd.read_stata('PS3_data.dta')
est_data.head(n=10)

id68  year  intid relhh  hannhrs  wannhrs  hlabinc  wlabinc  nochild  \
0     1  1967      1  Head   1200.0   2000.0      NaN      NaN        0   
1     2  1967      2  Head      0.0      0.0      NaN      NaN        0   
2     3  1967      3  Head      0.0      0.0      NaN      NaN        0   
3     4  1967      4  Head   1560.0      0.0      NaN      NaN        6   
4     5  1967      5  Head   2500.0   2000.0      NaN      NaN        3   
5     6  1967      6  Head   2230.0    120.0      NaN      NaN        2   
6     7  1967      7  Head   2254.0   1144.0      NaN      NaN        3   
7     8  1967      8  Head   1763.0   1960.0      NaN      NaN        2   
8     9  1967      9  Head      0.0      0.0      NaN      NaN        0   
9    10  1967     10  Head   1560.0      0.0      NaN      NaN        3   

   wrace  ...  redpregovinc  hsex  wsex   age  wage  hpersno  wpersno  hyrsed  \
0    NaN  ...        5614.0   1.0   2.0  52.0  46.0      1.0      2.0     8.0   
1    NaN  ...           0.0   1.0   2.0  56.0  57.0      1.0      2.0     3.0   
2    NaN  ...           0.0   1.0   2.0  77.0  64.0      1.0      2.0     NaN   
3    1.0  ...        3280.0   1.0   2.0  45.0  44.0      1.0      2.0     8.0   
4    1.0  ...        7900.0   1.0   2.0  24.0  22.0      1.0      2.0    10.0   
5    1.0  ...        6269.0   1.0   2.0  49.0  46.0      1.0      2.0     8.0   
6    1.0  ...        8743.0   1.0   2.0  51.0  42.0      1.0      2.0     7.0   
7    NaN  ...        9677.0   1.0   2.0  47.0  43.0      1.0      2.0     5.0   
8    NaN  ...           1.0   1.0   2.0  66.0  63.0      1.0      2.0    12.0   
9    NaN  ...        6301.0   1.0   2.0  57.0  58.0      1.0      2.0    16.0   

   wyrsed  pce  
0     8.0  0.0  
1     3.0  0.0  
2     3.0  0.0  
3     5.0  0.0  
4     9.0  0.0  
5    13.0  0.0  
6     7.0  0.0  
7     9.0  0.0  
8    12.0  0.0  
9    11.0  0.0  

[10 rows x 52 columns]

In [3]:
est_data.describe()

id68           year          intid        hannhrs  \
count  123786.000000  123786.000000  123786.000000  123786.000000   
mean     1494.639475    1984.831273    3271.379429    1679.269897   
std       838.901790       9.836212    2277.056058    1061.704712   
min         1.000000    1967.000000       1.000000       0.000000   
25%       772.000000    1977.000000    1444.000000     832.000000   
50%      1517.000000    1985.000000    2984.000000    1976.000000   
75%      2224.000000    1993.000000    4763.000000    2350.000000   
max      2930.000000    2002.000000   16968.000000    7800.000000   

             wannhrs       hlabinc        wlabinc        nochild  \
count  123786.000000  9.023300e+04   48496.000000  123786.000000   
mean      633.026917  4.211505e+04   22026.289062       0.843771   
std       878.422791  4.670424e+04   21336.107422       1.182829   
min         0.000000  6.353981e-01       1.192780       0.000000   
25%         0.000000  1.979858e+04    8016.247070       0.000000   
50%         0.000000  3.460022e+04   18122.412109       0.000000   
75%      1454.000000  5.267309e+04   30256.060547       2.000000   
max      5840.000000  3.771521e+06  856942.062500      11.000000   

              wrace          hrace  ...  redpregovinc           hsex     wsex  \
count  90603.000000  123656.000000  ...  1.237860e+05  123786.000000  80758.0   
mean       1.098220       1.129731  ...  3.012258e+04       1.233072      2.0   
std        0.356161       0.394627  ...  4.588795e+04       0.422940      0.0   
min        1.000000       1.000000  ... -1.324040e+05       1.000000      2.0   
25%        1.000000       1.000000  ...  7.700000e+03       1.000000      2.0   
50%        1.000000       1.000000  ...  1.900000e+04       1.000000      2.0   
75%        1.000000       1.000000  ...  3.910775e+04       1.000000      2.0   
max        8.000000       8.000000  ...  3.660000e+06       2.000000      2.0   

                 age          wage        hpersno       wpersno  \
count  123786.000000  80758.000000  123786.000000  80758.000000   
mean       45.545547     41.390785      39.620201     55.346169   
std        17.623671     14.786721      69.003265     77.864296   
min        16.000000     13.000000       1.000000      1.000000   
25%        31.000000     29.000000       1.000000      2.000000   
50%        42.000000     39.000000       3.000000      3.000000   
75%        58.000000     51.000000      22.000000    170.000000   
max       102.000000     95.000000     227.000000    231.000000   

              hyrsed        wyrsed            pce  
count  122809.000000  80091.000000  123786.000000  
mean       12.666091     12.720081       0.557690  
std         2.917721      2.422607       0.265198  
min         1.000000      1.000000       0.000000  
25%        12.000000     12.000000       0.362158  
50%        12.000000     12.000000       0.599887  
75%        15.000000     14.000000       0.786908  
max        17.000000     17.000000       0.928007  

[8 rows x 51 columns]

In [4]:
#Choose only those observations that have male heads of households, are between 25 and 60 years of age 
#and earn wages > $7/hr

est_data = est_data[(est_data['hannhrs'] > 0)]
est_data['hrlywage'] = est_data['hlabinc'] / est_data['hannhrs']
est_data = est_data[(est_data['hsex'] == 1) & (est_data['age'] > 25) & 
                    (est_data['age'] < 60) & (est_data['hrlywage'] > 7)]

In [5]:
#Create race dummy variables

est_data['black'] = (est_data['hrace'] == 2).astype(int)
est_data['hispanic'] = (est_data['hrace'] == 5).astype(int)
est_data['other'] = ((est_data['hrace'] == 3) | (est_data['hrace'] == 4) | (est_data['hrace'] == 6) | 
                     (est_data['hrace'] == 7)).astype(int)

In [6]:
#Create log of hourly wage (dependent variable)

est_data['lnwage'] = np.log(est_data['hrlywage'])

In [7]:
#Keep only the needed variables - log of wages, education, age, race dummies

est_data = est_data[['year','lnwage','hyrsed','age','black','hispanic','other']]

In [8]:
est_data.describe()

year        lnwage        hyrsed           age         black  \
count  54848.000000  54848.000000  54491.000000  54848.000000  54848.000000   
mean    1986.747757      3.020815     13.549283     39.478504      0.056082   
std        8.725407      0.543461      2.454004      9.142698      0.230083   
min     1971.000000      1.945946      1.000000     26.000000      0.000000   
25%     1980.000000      2.646286     12.000000     32.000000      0.000000   
50%     1987.000000      3.002805     13.000000     38.000000      0.000000   
75%     1994.000000      3.334669     16.000000     47.000000      0.000000   
max     2002.000000      7.448526     17.000000     59.000000      1.000000   

       hispanic         other  
count   54848.0  54848.000000  
mean        0.0      0.023046  
std         0.0      0.150049  
min         0.0      0.000000  
25%         0.0      0.000000  
50%         0.0      0.000000  
75%         0.0      0.000000  
max         0.0      1.000000

Since the dataset doesn't contain any Hispanic individuals, we do not use this variable in our optimization.

We now define the log-likelihood function that needs to be maximized in order to obtain the parameter estimates
of the model.

$$ l(\alpha, \beta) = \ln(\prod\limits_{i=1}^N p(y_i\mid x_i,\alpha,\beta) = \sum\limits_{i=1}^N \ln p(y_i\mid x_i,\alpha,\beta) $$

We assume that $\epsilon_i \sim N(0,\sigma^2)$. This leads to the likelihood function being of the form

\begin{equation}
\begin{split}
l(\alpha, \beta, \sigma^2) &=  \sum\limits_{i=1}^N \ln\left[\left(\frac{1}{2\pi\sigma^2}\right)^{1/2}\exp\left(-\frac{1}{2\pi\sigma^2}\left(y_i - \alpha - \beta^T x_i\right)^2\right)\right] \\
&= \frac{N}{2} \ln\left(\frac{1}{2\pi\sigma^2}\right) - \frac{1}{2\sigma^2}\sum\limits_{i=1}^N\left(y_i - \alpha - \beta^T x_i\right)^2
\end{split}
\end{equation}

We need to find $\alpha$, $\beta$ and $\sigma^2$ that maximize the above function. We begin by defining this function and then use a numerical optimization technique to find these maxima.

In [9]:
def obj_func(Beta, yearly_data):
    '''
    Compute the value of the log-likelihood function that is to be maximized
    
    Args:
        Beta: A length 6 tuple, model parameters (sigmasq,
               alpha, beta1, beta2, beta3, beta5)
        yearly_data: Data of the particular year for which the model is to be estimated
        
    Returns:
        ll_val: Value of the log-likelihood function
    '''
    
    sigmasq, alpha, beta1, beta2, beta3, beta5 = Beta
    
    SSE = ((yearly_data['lnwage'] - alpha - (beta1 * yearly_data['hyrsed'] + beta2 * yearly_data['age'] + 
                                          beta3 * yearly_data['black'] + beta5 * yearly_data['other'])) ** 2).sum()
    
    n = len(yearly_data)
    
    #Compute the log-likelihood
    #We use the negative log-likelihood since we're call a minimization routine
    
    ll_val = -((n / 2) * (np.log(1 / 2 * np.pi * sigmasq)) - (1 / 2 * sigmasq) * SSE)
    
    return ll_val

We now call `obj_func()` within an optimization routine. We use the Nelder-Mead method to solve the optimization problem since this is least sensitive to the starting values.

## Estimation for the year 1971

In [10]:
#Consider only the year 1971

data_71 = est_data[(est_data['year'] == 1971)]

#We use the correlation between variables as the initial values for the beta's,
#the variance of the dependent variable as the initial value for sigmasq, and
#the mean of the dependent variable as the initial value for alpha

sig_start = (data_71.loc[:,'lnwage'].std())**2
alpha_start = data_71.loc[:,'lnwage'].mean()

Beta0 = (sig_start, alpha_start, data_71['lnwage'].corr(data_71['hyrsed']),
         data_71['lnwage'].corr(data_71['age']), data_71['lnwage'].corr(data_71['black']),
         data_71['lnwage'].corr(data_71['other']))

# Use Nelder-Mead to minimize the negative of the log-likelihood function

min_results_71 = opt.minimize(obj_func, Beta0, args=(data_71),
                           method="Nelder-Mead", tol=1e-15)

print('The estimates for 1971 are:')
print('alpha = ', min_results_71['x'][1])
print('beta1 = ', min_results_71['x'][2])
print('beta2 = ', min_results_71['x'][3])
print('beta3 = ', min_results_71['x'][4])
print('beta5 = ', min_results_71['x'][5])

The estimates for 1971 are:
alpha =  1.612417161443112
beta1 =  0.0665156207958787
beta2 =  0.013311623129936556
beta3 =  -0.17970226267564604
beta5 =  -0.08764416688691612


## Interpretation of $\beta_1$ for the year 1971

The coefficient $\beta_1$ tells us that keeping all other variables constant, an additional year of education leads to a 6.9% increase in wages since $\exp(0.0665156207958787) = 1.06877766 \approx 1.069$

In [11]:
#We comapre the results of our MLE approach with OLS estimates

data_71['const'] = 1
var_list = ['const', 'hyrsed','age','black','other']
reg1 = sm.OLS(data_71['lnwage'], data_71[var_list], missing='drop')
results = reg1.fit()
results.params

/Users/Aditya/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


const     1.604855
hyrsed    0.066661
age       0.013387
black    -0.176535
other     0.005163
dtype: float64

In [12]:
print('The absolute difference between the estimates for alpha for 1971 is ',
      abs(min_results_71['x'][1] - results.params[0]))
print('The absolute difference between the estimates for beta1 for 1971 is ',
      abs(min_results_71['x'][2] - results.params[1]))
print('The absolute difference between the estimates for beta2 for 1971 is ',
      abs(min_results_71['x'][3] - results.params[2]))
print('The absolute difference between the estimates for beta3 for 1971 is ',
      abs(min_results_71['x'][4] - results.params[3]))
print('The absolute difference between the estimates for beta5 for 1971 is ',
      abs(min_results_71['x'][5] - results.params[4]))

The absolute difference between the estimates for alpha for 1971 is  0.0075616859213325505
The absolute difference between the estimates for beta1 for 1971 is  0.00014560577001961061
The absolute difference between the estimates for beta2 for 1971 is  7.569161163153634e-05
The absolute difference between the estimates for beta3 for 1971 is  0.003166790730705138
The absolute difference between the estimates for beta5 for 1971 is  0.09280706192915825


## Estimation for the year 1980

In [13]:
#Estimate the model for 1980

data_80 = est_data[(est_data['year'] == 1980)]

sig_start = (data_80.loc[:,'lnwage'].std())**2
alpha_start = data_80.loc[:,'lnwage'].mean()

Beta0 = (sig_start, alpha_start, data_80['lnwage'].corr(data_80['hyrsed']),
         data_80['lnwage'].corr(data_80['age']), data_80['lnwage'].corr(data_80['black']),
         data_80['lnwage'].corr(data_80['other']))

# Use Nelder-Mead to minimize the negative of the log-likelihood function

min_results_80 = opt.minimize(obj_func, Beta0, args=(data_80),
                           method="Nelder-Mead", tol=1e-15)

print('The estimates for 1980 are:')
print('alpha = ', min_results_80['x'][1])
print('beta1 = ', min_results_80['x'][2])
print('beta2 = ', min_results_80['x'][3])
print('beta3 = ', min_results_80['x'][4])
print('beta5 = ', min_results_80['x'][5])

The estimates for 1980 are:
alpha =  1.6173648234541051
beta1 =  0.06746473535251973
beta2 =  0.012830457130799375
beta3 =  -0.09180252051287374
beta5 =  -0.030741212289645735


## Interpretation of $\beta_1$ for the year 1980

The coefficient $\beta_1$ tells us that keeping all other variables constant, an additional year of education leads to a 7% increase in wages since $\exp(0.06746473535251973) = 1.06979253 \approx 1.07$

In [14]:
#We comapre the results of our MLE approach with OLS estimates

data_80['const'] = 1
reg2 = sm.OLS(data_80['lnwage'], data_80[var_list], missing='drop')
results2 = reg2.fit()
results2.params

/Users/Aditya/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


const     1.615687
hyrsed    0.067526
age       0.012823
black    -0.089979
other     0.019603
dtype: float64

In [15]:
print('The absolute difference between the estimates for alpha for 1980 is ',
      abs(min_results_80['x'][1] - results2.params[0]))
print('The absolute difference between the estimates for beta1 for 1980 is ',
      abs(min_results_80['x'][2] - results2.params[1]))
print('The absolute difference between the estimates for beta2 for 1980 is ',
      abs(min_results_80['x'][3] - results2.params[2]))
print('The absolute difference between the estimates for beta3 for 1980 is ',
      abs(min_results_80['x'][4] - results2.params[3]))
print('The absolute difference between the estimates for beta5 for 1980 is ',
      abs(min_results_80['x'][5] - results2.params[4]))

The absolute difference between the estimates for alpha for 1980 is  0.0016775594897435742
The absolute difference between the estimates for beta1 for 1980 is  6.15602653612074e-05
The absolute difference between the estimates for beta2 for 1980 is  7.157225769754277e-06
The absolute difference between the estimates for beta3 for 1980 is  0.0018232429831488567
The absolute difference between the estimates for beta5 for 1980 is  0.050344400094106005


## Estimation for the year 1990

In [16]:
#Estimate the model for 1990

data_90 = est_data[(est_data['year'] == 1990)]

sig_start = (data_90.loc[:,'lnwage'].std())**2
alpha_start = data_90.loc[:,'lnwage'].mean()

Beta0 = (sig_start, alpha_start, data_90['lnwage'].corr(data_90['hyrsed']),
         data_90['lnwage'].corr(data_90['age']), data_90['lnwage'].corr(data_90['black']),
         data_90['lnwage'].corr(data_90['other']))

# Use Nelder-Mead to minimize the negative of the log-likelihood function

min_results_90 = opt.minimize(obj_func, Beta0, args=(data_90),
                           method="Nelder-Mead", tol=1e-15)

print('The estimates for 1990 are:')
print('alpha = ', min_results_90['x'][1])
print('beta1 = ', min_results_90['x'][2])
print('beta2 = ', min_results_90['x'][3])
print('beta3 = ', min_results_90['x'][4])
print('beta5 = ', min_results_90['x'][5])

The estimates for 1990 are:
alpha =  1.0620897447681505
beta1 =  0.09938306728664142
beta2 =  0.014323874839676988
beta3 =  -0.1569130944969035
beta5 =  0.13630951904341004


## Interpretation of $\beta_1$ for the year 1990

The coefficient $\beta_1$ tells us that keeping all other variables constant, an additional year of education leads to a 10.5% increase in wages since $\exp(0.09938306728664142) = 1.10448931 \approx 1.105$

In [17]:
#We comapre the results of our MLE approach with OLS estimates

data_90['const'] = 1
reg3 = sm.OLS(data_90['lnwage'], data_90[var_list], missing='drop')
results3 = reg3.fit()
results3.params

/Users/Aditya/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


const     1.065945
hyrsed    0.099325
age       0.014340
black    -0.167100
other    -0.066803
dtype: float64

In [18]:
print('The absolute difference between the estimates for alpha for 1990 is ',
      abs(min_results_90['x'][1] - results3.params[0]))
print('The absolute difference between the estimates for beta1 for 1990 is ',
      abs(min_results_90['x'][2] - results3.params[1]))
print('The absolute difference between the estimates for beta2 for 1990 is ',
      abs(min_results_90['x'][3] - results3.params[2]))
print('The absolute difference between the estimates for beta3 for 1990 is ',
      abs(min_results_90['x'][4] - results3.params[3]))
print('The absolute difference between the estimates for beta5 for 1990 is ',
      abs(min_results_90['x'][5] - results3.params[4]))

The absolute difference between the estimates for alpha for 1990 is  0.0038551850672583488
The absolute difference between the estimates for beta1 for 1990 is  5.777823073087329e-05
The absolute difference between the estimates for beta2 for 1990 is  1.5735850456621153e-05
The absolute difference between the estimates for beta3 for 1990 is  0.010187158370672494
The absolute difference between the estimates for beta5 for 1990 is  0.20311298274688072


## Estimation for the year 2000

In [19]:
#Estimate the model for 2000

data_00 = est_data[(est_data['year'] == 2000)]

sig_start = (data_00.loc[:,'lnwage'].std())**2
alpha_start = data_00.loc[:,'lnwage'].mean()

Beta0 = (sig_start, alpha_start, data_00['lnwage'].corr(data_00['hyrsed']),
         data_00['lnwage'].corr(data_00['age']), data_00['lnwage'].corr(data_00['black']),
         data_00['lnwage'].corr(data_00['other']))

# Use Nelder-Mead to minimize the negative of the log-likelihood function

min_results_00 = opt.minimize(obj_func, Beta0, args=(data_00),
                           method="Nelder-Mead", tol=1e-15)

print('The estimates for 2000 are:')
print('alpha = ', min_results_00['x'][1])
print('beta1 = ', min_results_00['x'][2])
print('beta2 = ', min_results_00['x'][3])
print('beta3 = ', min_results_00['x'][4])
print('beta5 = ', min_results_00['x'][5])

The estimates for 2000 are:
alpha =  1.1336944718452626
beta1 =  0.11062249356211762
beta2 =  0.01099265162280777
beta3 =  -0.1783291221427089
beta5 =  0.1818305903108428


## Interpretation of $\beta_1$ for the year 2000

The coefficient $\beta_1$ tells us that keeping all other variables constant, an additional year of education leads to a 11.7% increase in wages since $\exp(0.11062249356211762) = 1.11697316 \approx 1.117$

In [20]:
#We comapre the results of our MLE approach with OLS estimates

data_00['const'] = 1
reg4 = sm.OLS(data_00['lnwage'], data_00[var_list], missing='drop')
results4 = reg4.fit()
results4.params

/Users/Aditya/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


const     1.163822
hyrsed    0.109950
age       0.010779
black    -0.253212
other    -0.055265
dtype: float64

In [21]:
print('The absolute difference between the estimates for alpha for 2000 is ',
      abs(min_results_00['x'][1] - results4.params[0]))
print('The absolute difference between the estimates for beta1 for 2000 is ',
      abs(min_results_00['x'][2] - results4.params[1]))
print('The absolute difference between the estimates for beta2 for 2000 is ',
      abs(min_results_00['x'][3] - results4.params[2]))
print('The absolute difference between the estimates for beta3 for 2000 is ',
      abs(min_results_00['x'][4] - results4.params[3]))
print('The absolute difference between the estimates for beta5 for 2000 is ',
      abs(min_results_00['x'][5] - results4.params[4]))

The absolute difference between the estimates for alpha for 2000 is  0.03012767593660115
The absolute difference between the estimates for beta1 for 2000 is  0.0006729433923380601
The absolute difference between the estimates for beta2 for 2000 is  0.0002136403613697043
The absolute difference between the estimates for beta3 for 2000 is  0.07488323230676291
The absolute difference between the estimates for beta5 for 2000 is  0.2370957364512935


## Comments on MLE using a numerical optimization routine

We see that maximizing the likelihood function using Nelder-Mead provides estimates that are reasonably close to the OLS estimates. The estimate for $\beta_5$ is often considerably different from the OLS estimate, but the others are very close. Using a gradient based method like BFGS leads to poor results unless the initial points fed to the optimization routine are very close to the optima.

## The change in returns to education

We see that the coefficient $\beta_1$ is increasing over the four different models. This suggests that an additional year of education provides a larger return (higher wages) over subsequent decades.